### PR TITLE
Allow custom HTML in `<head>`.

### PR DESCRIPTION
**Pull request by vizziv**
Adds `headHtml : Optional Text` option to the configuration. It can be set in one of two ways, but not both simultaneously.

- The option can be set as usual in `./neuron.dhall`.
- If the file `./head.html` exists, then `headHtml` is set to `Some [head.html's contents]`.

Addresses #362 directly. Also solves or enables workarounds for several other issues (e.g. #212, #361, #383).

### CODE REVIEW

**Comment by srid** [comment]
@vizziv Thanks! Taking a quick look at the diff, I see that this differs from #362 in two ways:

- Why the neuron.dhall configuration, over just using `head.html` if it exists (see [convention over configuration](https://en.wikipedia.org/wiki/Convention_over_configuration))?
- Why is the default mathjax script still included? I thought we would have a default snippet for `head.html` when it is not specified by the user, and that this snippet would contain the existing mathjax include. The idea being that if the user overrides head.html, they become responsible for including whatever math script (math or katex or whatever) they need.

**Comment by vizziv** [comment]
The answer in both cases is that I chose a solution that made the fewest changes to the code, as I wanted to submit the minimally invasive patch that added the feature. But I'm happy to make both changes.

Regarding how to organize the code for the first change: should we use the `Config` type to store `head.html`'s contents, or should we create a new module and type to handle it? (Or should we just put the code reading `head.html` directly into `Main`?)

**Comment by vizziv** [comment]
Another question about the second change: right now we replace `$`/`$` pairs with `\(`/`\)` pairs for MathJax's sake. Should disabling MathJax entail disabling this replacement?

**Comment by srid** [comment]
Checkout how the [icon and manifest files](https://neuron.zettel.page/67c6f7a0.html) are handled for inspiration. They are loaded in src/app/Main.hs as `manifest` and passed around (along with `config`) as a function argument to `renderRoutePage`. You can do the same for `head.html`. And notice how `Neuron.Web.Manifest` handles the creation of the manifest type, including handing the default case - of using neuron logo as the favicon.

**Comment by srid** [comment]
re: MathJax - I'm not sure what you mean; neuron doesn't do any special processing, aside from letting commonmark handle it (see `CE.mathSpec` in `Neuron.Reader.Markdown`). The `mathJaxSupport` config field should be ... made to be totally ignored in this PR (with a deprecation warning), and entirely removed one day, I think.

**Comment by srid** [comment]
You can probably sneak in the head.html stuff in a new field of the `Neuron.Web.Manifest:Manifest` type, while renaming that module/type suitably (it won't just be a "manifest" anymore).

**Comment by vizziv** [comment]
Ahh, I'd already made a new module for this, creatively named `HeadHtml`. But it could definitely be easily combined with (an appropriately renamed) `Manifest` either now or in the future.

**Comment by vizziv** [comment]
> The `mathJaxSupport` config field should be ... made to be totally ignored in this PR (with a deprecation warning), and entirely removed one day, I think.

I've removed `mathJaxSupport` from `Config`. I wasn't sure how to best add a deprecation warning, so instead we throw an error. This also throws an error for other spurious fields in `neuron.dhall`.

I'm not sure how you want to update the guide to reflect these changes, so I'll leave that to you and others if that's okay.

> re: MathJax - I'm not sure what you mean; neuron doesn't do any special processing, aside from letting commonmark handle it (see `CE.mathSpec` in `Neuron.Reader.Markdown`).

Yeah, looks like the `$`/`$`-to-`\(`/`\)` conversion is determined by some combination of `commonmark-pandoc` and `pandoc`. Will table this for now.

File: neuron/src/app/Main.hs
Comment (srid) [inline neuron/src/app/Main.hs:2]: This can be removed.

File: neuron/src/app/Main.hs
Comment (srid) [inline neuron/src/app/Main.hs:44]: `config` should come before `headHtml` in arguments.

File: neuron/src/app/Neuron/Config.hs
Comment (srid) [inline neuron/src/app/Neuron/Config.hs:73]: 1. There should be a comment explaining what this piece of code does.

1. If spuriousFields is a list of length 1 containing just "mathJaxSupport" - it should be ignored, without returning an error. For other cases, it should error out. The idea is that we remove `mathJaxSupport` entirely, while still allowing its presence in older neuron.dhall (if only to ignore it). Running the new neuron on an older zettelkasten shouldn't fail.

File: neuron/src/app/Neuron/Config.hs
Comment (srid) [inline neuron/src/app/Neuron/Config.hs:46]: Hmm, what's the reason for this change? Default config is not used anymore?

File: neuron/src/app/Neuron/Web/View.hs
Comment (srid) [inline neuron/src/app/Neuron/Web/View.hs:58]: `config` should come before `headHtml` in arguments.

File: neuron/src/app/Neuron/Web/View.hs
Comment (srid) [inline neuron/src/app/Neuron/Web/View.hs:92]: I suggest moving the default HTML to `HeadHtml.hs`, and thereby changing `renderHtmlOr headHtml (elAttr ...)` to `elPandocRaw (Format "html") headHtml`. Why? Because that's what we do with `Manifest` (the default is handled in the module itself).

File: neuron/src/lib/Neuron/Config/Type.hs
Comment (srid) [inline neuron/src/lib/Neuron/Config/Type.hs:13]: ?

**Review by srid — CHANGES_REQUESTED** [review]

**Comment by srid** [comment]
About the deprecation warning .. okay, let's not worry about it; let's ignore mathjax silently if used by the user. I added some comments in your PR.

About the $ -> ( conversion, that happens here: https://github.com/srid/reflex-dom-pandoc/blob/259c7a881dd4ba69f82ac2fdec83ddf8f5ee0fc5/src/Reflex/Dom/Pandoc/Document.hs#L204

File: neuron/src/app/Neuron/Config.hs
Comment (vizziv) [reply neuron/src/app/Neuron/Config.hs:46]: Ahh, the introduction of `emptyConfig` was more important in a previous iteration of the code. We can remove it.

With that said, the `Config` interface was a little strange before. Exporting both `defaultConfig` and `mergeWithDefault` felt like a bit of a code smell. And on the line after this one, we would sometimes end up merging two copies of `defaultConfig`.

Proposed solution: change the `Config` interface to export neither `emptyConfig` nor `defaultConfig`, but change `mergeWithDefault` to
```haskell
mergeWithDefault :: Maybe Text -> Text
mergeWithDefault = \case
  Nothing -> defaultConfig
  Just config -> defaultConfig <> " // " <> config
```
Then this line of code can be `pure Nothing`.

File: neuron/src/app/Neuron/Config.hs
Comment (vizziv) [reply neuron/src/app/Neuron/Config.hs:73]: Okay, I'll add an explanation and a special case for when `mathJaxSupport` is the only spurious field.

(Also, the "TODO: remove once" comment was meant to continue: ... "mathJaxSupport has been deprecated for long enough")

File: neuron/src/lib/Neuron/Config/Type.hs
Comment (vizziv) [reply neuron/src/lib/Neuron/Config/Type.hs:13]: See previous comment about `Config.Type`.

File: neuron/src/app/Neuron/Config.hs
Comment (srid) [reply neuron/src/app/Neuron/Config.hs:46]: Sounds good, re: proposed solution.

File: neuron/src/app/Neuron/Config.hs
Comment (srid) [inline neuron/src/app/Neuron/Config.hs:79]: Config parsing has gotten a bit complex. I'm wondering if we should just break backwards compact, and have users remove this field manually.

File: neuron/src/app/Neuron/Web/View.hs
Comment (srid) [inline neuron/src/app/Neuron/Web/View.hs:55]: ?

**Review by srid — CHANGES_REQUESTED** [review]
@vizziv You know what, I'm now not sure if it is so important to support backwards compact, especially if it makes config parsing rather complex. 

Most users I imagine won't have `mathJaxSupport = false` in their config. Those that do, can always remove it when next updating neuron.

So let's get rid of the config fields check entirely. Please undo the Config changes in your PR, except for removing `mathJaxSupport` field entirely from the Config type.

File: neuron/src/app/Neuron/Config.hs
Comment (srid) [reply neuron/src/app/Neuron/Config.hs]: No need to make this change in this PR (see comment below).

File: neuron/src/app/Neuron/Web/View.hs
Comment (srid) [reply neuron/src/app/Neuron/Web/View.hs:55]: btw, you can use `blank` in place of `pure ()`.

Comment (vizziv) [reply neuron/src/app/Neuron/Web/View.hs:55]: Oh, whoops, I had forgotten about this DOCTYPE addition. Handling DOCTYPE could be its own PR, in which case I'll remove this, but I also suspect that `<!DOCTYPE html>` is the right thing for most people.

The reasoning for adding it was for loading KaTeX, which doesn't work in quirks mode. One case for including the change in this PR is that #212 is one of this PR's motivations.

Comment (srid) [reply neuron/src/app/Neuron/Web/View.hs:55]: Okay, feel free to leave it in then - but do add that note (re: KaTeX) as a code comment above that.

**Review by srid — APPROVED** [review]

**Comment by vizziv** [comment]
Okay, I undid the config changes (except I removed the `mathJaxSupport` field) and explained DOCTYPE.

The `Neuron.Config` and `Neuron.Config.Type` commit histories on this branch are now kind of messy, so squashing these commits upon merging makes sense.

**Pull request merged**
